File: frontend/src/app/core/models/qa.ts
export interface EvidenceTriple {
  subject_id: string;
  subject_label: string;
  predicate: string;
  object_id?: string | null;
  object_label?: string | null;
  object_value?: string | null;
}

export interface Evidence {
  id: string;
  kind: string;
  entity_id?: string | null;
  label: string;
  class_label?: string | null;
  properties?: Record<string, unknown>;
  triples?: EvidenceTriple[];
  source_ref?: Record<string, unknown> | null;
}

export interface QaMessage {
  id: string;
  role: 'user' | 'assistant' | string;
  content: string;
  evidences?: Evidence[];
  plan?: Record<string, unknown> | null;
  tool_trace?: Array<Record<string, unknown>>;
  created_at?: string | null;
}

export interface QaSessionSummary {
  id: string;
  ontology_model_id: string;
  llm_model_id?: string | null;
  title: string;
  message_count: number;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface QaSession {
  id: string;
  ontology_model_id: string;
  llm_model_id?: string | null;
  title: string;
  resolved_entities?: Record<string, unknown> | null;
  created_at?: string | null;
  updated_at?: string | null;
  messages: QaMessage[];
}

export interface QaChatResponse {
  session_id: string;
  answer: string;
  evidences: Evidence[];
  plan?: Record<string, unknown> | null;
  tool_trace?: Array<Record<string, unknown>>;
  resolved_entities?: Record<string, unknown> | null;
}

export interface KnowledgeAccessLog {
  id: string;
  created_at?: string | null;
  caller: string;
  tool_name: string;
  ontology_model_id?: string | null;
  session_id?: string | null;
  trace_id: string;
  latency_ms: number;
  empty_hit: boolean;
  error?: string | null;
}

File: backend/app/schemas/qa.py
"""QA session and chat DTOs."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.knowledge.evidence import Evidence


class QaSessionCreate(BaseModel):
    ontology_model_id: str
    model_id: str | None = None


class QaSessionUpdate(BaseModel):
    title: str = Field(min_length=1, max_length=80)


class QaSessionSummary(BaseModel):
    id: str
    ontology_model_id: str
    llm_model_id: str | None = None
    title: str = ""
    message_count: int = 0
    created_at: datetime | None = None
    updated_at: datetime | None = None


class QaMessageRead(BaseModel):
    id: str
    role: str
    content: str
    evidences: list[Evidence] = Field(default_factory=list)
    plan: dict[str, Any] | None = None
    tool_trace: list[dict[str, Any]] = Field(default_factory=list)
    created_at: datetime | None = None


class QaSessionRead(BaseModel):
    id: str
    ontology_model_id: str
    llm_model_id: str | None = None
    title: str = ""
    resolved_entities: dict[str, Any] | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    messages: list[QaMessageRead] = Field(default_factory=list)


class QaChatRequest(BaseModel):
    question: str = Field(min_length=1, max_length=4000)
    model_id: str | None = None


class QaChatResponse(BaseModel):
    session_id: str
    answer: str
    evidences: list[Evidence] = Field(default_factory=list)
    plan: dict[str, Any] | None = None
    tool_trace: list[dict[str, Any]] = Field(default_factory=list)
    resolved_entities: dict[str, Any] | None = None
